### PR TITLE
[WebProfiler] Improve HttpClient Panel

### PR DESCRIPTION
**Pull request by ismail1432**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no    
| Deprecations? | no 
| Tests pass?   | np    
| Fixed tickets | #33311  
| License       | MIT
| Doc PR        | ?

Wishlist from issue :

- [x] the panel UI could be improved I suppose, e.g. a tooltip on hover in the toolbar
- [ ] putting requests in the "Performance" tab would be great (all timings are already in the collected "info")
- [ ] ~response bodies are not collected yet~ https://github.com/symfony/symfony/pull/33315#discussion_r317353605
- [ ] when a transport error occurs, (ie the "error" info is populated) we should maybe have a better display too?
- [x] add a copy-as-curl button next to each request

**[EDIT 07 oct 19]**
This PR contains :
- Display a tooltip in Toolbar
- Add a link only for GET Request in profiler panel
image

### CODE REVIEW

**Comment by ismail1432** [comment]
I'll take this one if it's possible but I have few questions :

> putting requests in the "Performance" tab would be great (all timings are already in the collected "info")

Does display the total of request in **Performance metrics** section is enough (see picture) or we should also implement details for each request in **Execution timeline** section

image

> add a copy-as-curl button next to each request

I just add a href link, I don't know how to implement a copy as curl button in pure html but the user can still copy the link :-)

image

> response bodies are not collected yet

Body is returned via `toArray` IMO it's more readable and it's display a caret thanks to VarDumper :heart:

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/http_client.html.twig [hunk 1]
```diff
@@ -8,6 +8,17 @@
             <span class="sf-toolbar-value">{{ collector.requestCount }}</span>
         {% endset %}
 
+        {% set text %}
+            <div class="sf-toolbar-info-piece">
+                <b>Total Requests</b>
```
Comment (ro0NL) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/http_client.html.twig]: `Total requests`

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/http_client.html.twig [hunk 2]
```diff
@@ -8,6 +8,17 @@
...
+                <span>{{ collector.requestCount }}</span>
+            </div>
+            <div class="sf-toolbar-info-piece">
+                <b>HTTP errors</b>
+                <span>{{ collector.errorCount }}</span>
```
Comment (ro0NL) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/http_client.html.twig]: `class="sf-toolbar-status sf-toolbar-status-red"`

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/http_client.html.twig [hunk 3]
```diff
@@ -61,7 +72,7 @@
                                             <span class="label">{{ trace.method }}</span>
                                         </th>
                                         <th class="full-width">
-                                            {{ trace.url }}
+                                            <a href="{{ trace.url }}">{{ trace.url }}</a>
```
Comment (ro0NL) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/http_client.html.twig]: :+1:

File: src/Symfony/Component/HttpClient/TraceableHttpClient.php [hunk 4]
```diff
@@ -50,7 +51,10 @@ public function request(string $method, string $url, array $options = []): Respo
             }
         };
 
-        return $this->client->request($method, $url, $options);
+        $response = $this->client->request($method, $url, $options);
+        $body = $response->toArray();
```
Comment (ro0NL) [inline src/Symfony/Component/HttpClient/TraceableHttpClient.php]: this defeats lazyness i believe ... shouldnt we show the body "if requested" during the app's lifecycle? If the request was never made, the profiler should display it as such IMHO and not perform the actual request.

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/TraceableHttpClient.php]: Correct. This is not easy to achieve with the current available infrastructure. I'd suggest postponing this part for later.

Comment (ismail1432) [reply src/Symfony/Component/HttpClient/TraceableHttpClient.php]: Ok I removed this part from PR/Issue but I'm curious to know what's your idea or the expected behavior ? A button that display (= retrieve) body only if clicked

**Comment by nicolas-grekas** [comment]
Now I'm wondering: where are the request options, which contain the request headers? It might be useful to display them too.

File: src/Symfony/Component/HttpClient/TraceableHttpClient.php (hunk 4 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/TraceableHttpClient.php]: The body should be caught while it is being downloaded, then just displayed here. Monitoring downloads requires some code infrastructure that's not available for now.

**Comment by nicolas-grekas** [comment]
For copy-as-curl, we should be able to borrow from the copy/paste buttons here on the GitHub UI for example.
Then, it's not only about a link, but about all headers. See copy-as-curl from the browser network panel.

**Comment by ismail1432** [comment]
image

Thanks for reaction :+1:  
Now we display request headers (see picture).
I tried without success to get the Response while downloading :disappointed: 
IMO Response body, headers and info should be displayed into a table too instead of being returned as a dump. This is why I removed the dump at this time.

Few questions : 
Should we display all request options ? (for me Yes)
We can't implement without pain a copy-as-curl functionality in vanilla JS, can we use third JS library (exemple : https://www.npmjs.com/package/request-to-curl) ?

**Comment by nicolas-grekas** [comment]
I personally prefer a compact layout, so that when 10+ requests are made, you don't have to scroll 10+ screens to seem them all. But I'll let others decide what's best as design is not my 1st skill...

**Comment by stof** [comment]
another solution for the case of many requests is to allow collapsing each request (we already have the JS for that)

**Comment by nicolas-grekas** [comment]
@ismail1432 up to move this one forward? If there are things that you don't have time to finish, please remove them from the PR so we can merge the ready parts.

**Comment by ismail1432** [comment]
Sorry for the delay, I'll remove asap the part that I can't finish.

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/http_client.html.twig (hunk 3 above)
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/http_client.html.twig]: I would make it a link only when the method is GET.

Comment (ismail1432) [reply src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/http_client.html.twig]: shoudl be done

**Comment by ismail1432** [comment]
done @nicolas-grekas 
I update the description with the content of the `PR`

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/http_client.html.twig (hunk 3 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/http_client.html.twig]: Can you please revert this change?
A copy-as-curl button would be a better alternative. Meanwhile, no need to add this link.

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/http_client.html.twig [hunk 5]
```diff
@@ -8,6 +8,17 @@
...
+        {% set text %}
+            <div class="sf-toolbar-info-piece">
+                <b>Total requests</b>
+                <span>{{ collector.requestCount }}</span>
+            </div>
+            <div class="sf-toolbar-info-piece">
+                <b>HTTP errors</b>
+                <span class="sf-toolbar-status sf-toolbar-status-red">{{ collector.errorCount }}</span>
+            </div>
+        {% endset %}
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/http_client.html.twig]: can you please share a screenshot showing the effect of this?

Comment (nicolas-grekas) [reply src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/http_client.html.twig]: Oh, OK already here:
image

**Comment by nicolas-grekas** [comment]
Thank you @ismail1432.

**Pull request merged**